airflow/models/base_async_operator.py

```diff
@@ -0,0 +1,161 @@
+        return None
+
+      def poke(self, context: Dict) -> bool:
+        return bool(random.getrandbits(1))
+    ```
+
+    AsyncOperators must override the following methods:
+    :py:meth:`submit_request`: fire a request for a long running operation
+    :py:meth:`poke`: a method to check if the long running operation is
+    complete it should return True when a success criteria is met.
+
+    Optionally, AsyncOperators can override:
+    :py:meth: `process_result` to perform any operations after the success
+    criteria is met in :py:meth: `poke`
+
+    :py:meth: `poke` is executed at a time interval and succeed when a
+    criteria is met and fail if and when they time out.
+
+    :param soft_fail: Set to true to mark the task as SKIPPED on failure
+    :type soft_fail: bool
+    :param poke_interval: Time in seconds that the job should wait in
+        between each tries
+    :type poke_interval: int
+    :param timeout: Time, in seconds before the task times out and fails.
+    :type timeout: int
+
+    """
+    ui_color = '#9933ff'  # type: str
+
+    @apply_defaults
+    def __init__(self,
+                 *args,
+                 **kwargs) -> None:
+        super().__init__(mode='reschedule', *args, **kwargs)
+
+    @abstractmethod
+    def submit_request(self, context: Dict) -> Optional[Union[str, List, Dict]]:
+        """
+        This method should kick off a long running operation.
+        This method should return the ID for the long running operation if
+        applicable.
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+
+        :returns: a resource_id for the long running operation.
+        :rtype: Optional[Union[String, List, Dict]]
+        """
+        raise NotImplementedError
+
+    def process_result(self, context: Dict):
+        """
+        This method can optionally be overriden to process the result of a long running operation.
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+        """
+        self.log.info('Using default process_result. Got result of %s. Done.',
+                      self.get_external_resource_id(context))
+
+    def execute(self, context: Dict) -> None:
+        # On the first execute call submit_request and set the
+        # external resource id.
+
+        # pylint: disable=no-value-for-parameter
+        task_reschedules = TaskReschedule.find_for_task_instance(context['ti'])
+        if not task_reschedules:
+            resource_id = self.submit_request(context)
+            self.set_external_resource_id(context, resource_id)
+
+        super().execute(context)
+
+        resource_id = self.get_external_resource_id(context)
+        self.log.info("Calling process_result for %s.", resource_id)
+        self.process_result(context)
+        self.set_external_resource_id(context, None)
```

Comment:
If the class retained the XCom value, it could potentially allow reuse of resources across multiple chained tasks? For example, say I wrote EmrRunStepsOperator as some composed replacement for the existing EmrCreateJobFlowOperator + EmrAddStepsOperator + EmrStepSensor and it extended your BaseAsyncOperator. It would need to take an EMR cluster specification as a constructor parameter to do the same as EmrCreateJobFlowOperator. But it could also take a parameter that's an existing EMR cluster id, which could be templated to allow the user to use XCom to fetch it from the previous task: if the cluster with that id already exists, reuse it, otherwise create a new cluster.

```
a = EmrRunStepsOperator(task_id='phase_1', job_flow_spec, cluster_id=None, steps=[ ... ])
b = EmrRunStepsOperator(task_id='phase_2', job_flow_spec,
    cluster_id="{{ task_instance.xcom_pull('phase_1', key=XCOM_EXTERNAL_RESOURCE_ID_KEY) }}",
    steps=[ ... ])
a >> b
```

The problem this solves is that it can be quite expensive to start EMR clusters, typically a few minutes, especially if they're based on "spot" AWS EC2 instances. I'm sure there are other resources that can be expensive to start. We could explicitly start the cluster with a separate task at the beginning of the chain, but when using spot instances these can also terminate prematurely so it's pretty desirable to wrap everything - including optional cluster creation - in the same operator for retry purposes.

Reply:
Yea i also like the idea of preserving the xcom value.  

I think people know or should know that xcom is one of many things in database you want to be mindful of with regard to purging and archiving as part of cluster maintenance.

So I would vote to not purge by default.